lib/xml_print.ml

```diff
@@ -48,16 +48,27 @@ let encode_unsafe_char_and_at s =
   Buffer.contents b
 
 let compose_decl ?(version = "1.0") ?(encoding = "UTF-8") () =
-  "<?xml version=\"" ^ version ^ "\" encoding=\"" ^ encoding ^ "\"?>\n"
+  Format.sprintf
+    {|<?xml version="%s" encoding="%s"?>\n|}
+    version encoding
 
 let compose_doctype dt args =
-  "<!DOCTYPE " ^ dt
-  ^ (if args = []
-     then ""
-     else
-       " PUBLIC " ^
-       String.concat " " (List.map (fun a -> "\"" ^ a ^ "\"") args)) ^ ">"
+  let pp_args fmt = function
+    | [] -> ()
+    | l ->
+      Format.fprintf fmt " PUBLIC %a"
+        (Format.pp_print_list ~pp_sep:Format.pp_print_space
+           (fun fmt -> Format.fprintf fmt "%S"))
+        l
+  in
+  Format.asprintf
+    "<!DOCTYPE %s%a>"
+    dt
+    pp_args args
 
+let re_end_comment = Re.(compile @@ str "-->")
+let escape_comment =
+  Re.replace_string ~all:true re_end_comment ~by:"--&gt;"
```

Comment:
In HTML (and IIRC XML), even `--` is not allowed in comments, so it may be more sensible to replace `-+` (as a regexp) with `-`. Also, the replacement `&gt;` doesn't have any special meaning in comments.